sdks/go/pkg/beam/registration/emitterIterRegistration.go

```diff
@@ -0,0 +1,298 @@
+// of the function, aka:
+// registration.RegisterEmitter2[T1, T2]((*func(T1, T2))(nil))
+func RegisterEmitter2[T1, T2 any](e *func(T1, T2)) {
+	registerFunc := func(n exec.ElementProcessor) exec.ReusableEmitter {
+		gen := &emitNative2[T1, T2]{n: n}
+		gen.fn = gen.invoke
+		return gen
+	}
+	if reflect.TypeOf(e).Elem().In(0) == typex.EventTimeType {
+		registerFunc = func(n exec.ElementProcessor) exec.ReusableEmitter {
+			gen := &emitNative1WithTimestamp[T2]{n: n}
+			gen.fn = gen.invoke
+			return gen
+		}
+	}
+	exec.RegisterEmitter(reflect.TypeOf(e).Elem(), registerFunc)
+}
+
+// RegisterEmitter3 registers parameters from your DoFn with a
+// signature func(T1, T2, T3) and optimizes their execution.
+// This must be done by passing in a reference to an instantiated version
+// of the function, aka:
+// registration.RegisterEmitter3[T1, T2, T3]((*func(T1, T2, T3))(nil))
+func RegisterEmitter3[T1, T2, T3 any](e *func(T1, T2, T3)) {
+	registerFunc := func(n exec.ElementProcessor) exec.ReusableEmitter {
+		gen := &emitNative2WithTimestamp[T2, T3]{n: n}
+		gen.fn = gen.invoke
+		return gen
+	}
+	exec.RegisterEmitter(reflect.TypeOf(e).Elem(), registerFunc)
+}
+
+type iterNative1[T any] struct {
+	s  exec.ReStream
+	fn interface{}
+
+	// cur is the "current" stream, if any.
+	cur exec.Stream
+}
+
+func (v *iterNative1[T]) Init() error {
+	cur, err := v.s.Open()
+	if err != nil {
+		return err
+	}
+	v.cur = cur
+	return nil
+}
+
+func (v *iterNative1[T]) Value() interface{} {
+	return v.fn
+}
+
+func (v *iterNative1[T]) Reset() error {
+	if err := v.cur.Close(); err != nil {
+		return err
+	}
+	v.cur = nil
+	return nil
+}
+
+func (v *iterNative1[T]) invoke(value *T) bool {
+	elm, err := v.cur.Read()
+	if err != nil {
+		if err == io.EOF {
+			return false
+		}
+		panic(fmt.Sprintf("broken stream: %v", err))
+	}
+	*value = elm.Elm.(T)
+	return true
+}
+
+type iterNative2[T1, T2 any] struct {
+	s  exec.ReStream
+	fn interface{}
+
+	// cur is the "current" stream, if any.
+	cur exec.Stream
+}
+
+func (v *iterNative2[T1, T2]) Init() error {
+	cur, err := v.s.Open()
+	if err != nil {
+		return err
+	}
+	v.cur = cur
+	return nil
+}
+
+func (v *iterNative2[T1, T2]) Value() interface{} {
+	return v.fn
+}
+
+func (v *iterNative2[T1, T2]) Reset() error {
+	if err := v.cur.Close(); err != nil {
+		return err
+	}
+	v.cur = nil
+	return nil
+}
+
+func (v *iterNative2[T1, T2]) invoke(key *T1, value *T2) bool {
+	elm, err := v.cur.Read()
+	if err != nil {
+		if err == io.EOF {
+			return false
+		}
+		panic(fmt.Sprintf("broken stream: %v", err))
+	}
+	*key = elm.Elm.(T1)
+	*value = elm.Elm2.(T2)
+	return true
+}
+
+// RegisterIter1 registers parameters from your DoFn with a
+// signature func(*T) bool and optimizes their execution.
+// This must be done by passing in a reference to an instantiated version
+// of the function, aka:
+// registration.RegisterIter1[T]((*func(*T) bool)(nil))
+func RegisterIter1[T any](i *func(*T) bool) {
```

Comment:
We may consider adding something for Map Side Input functions `func(K) func(*V) bool` here and "re iterators" `func() func(*V) bool` (and the KV version).

That's not critical for this pass, since we don't have the registration scaffolding for that set up, and I don't think there's a great need for it, because those typically have high RPC overhead anyway as they fetch & refetch the data.

We can always benchmark and add them later if needed.

The reflective versions are here: https://github.com/apache/beam/blob/51067c123204a4951de3cba858edc96da354292b/sdks/go/pkg/beam/core/runtime/exec/input.go#L71

Reply:
Yeah, I think its a good idea to do that eventually, I'd scope that out of this PR though. It shouldn't be too hard to build off of this one once we have a pattern established though.